server/service/async/async.go

```diff
@@ -0,0 +1,280 @@
+package async
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/fleetdm/fleet/v4/server/datastore/redis"
+	"github.com/fleetdm/fleet/v4/server/fleet"
+	kitlog "github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	redigo "github.com/gomodule/redigo/redis"
+	"github.com/jmoiron/sqlx"
+	"github.com/pkg/errors"
+)
+
+const (
+	labelMembershipHostKeyPattern = "label_membership:{*}"
+	labelMembershipHostKey        = "label_membership:{%d}"
+	labelMembershipReportedKey    = "label_membership_reported:{%d}"
+	collectorLockKey              = "locks:async_collector:{%s}"
+)
+
+type Task struct {
+	Datastore fleet.Datastore
+	Pool      fleet.RedisPool
+	// AsyncEnabled indicates if async processing is enabled in the
+	// configuration. Note that Pool can be nil if this is false.
+	AsyncEnabled bool // TODO: should this be read in a different way, more dynamically, if config changes while fleet is running? Or does that require a restart?
```

Comment:
^ Would like to clarify this, I think it's fine to be static and a restart is required if config changes, just want to make sure.